### PR TITLE
Add ast-grep recipe.

### PR DESCRIPTION
**Pull request by asford**
Checklist
- [x] Title of this PR is meaningful: e.g. "Adding my_nifty_package", not "updated meta.yaml".
- [x] License file is packaged (see [here](https://github.com/conda-forge/staged-recipes/blob/5eddbd7fc9d1502169089da06c3688d9759be978/recipes/example/meta.yaml#L64-L73) for an example).
- [x] Source is from official source.
- [x] Package does not vendor other packages. (If a package uses the source of another package, they should be separate packages or the licenses of all packages need to be packaged).
- [x] If static libraries are linked in, the license of the static library is packaged.
- [x] Package does not ship static libraries. If static libraries are needed, [follow CFEP-18](https://github.com/conda-forge/cfep/blob/main/cfep-18.md).
- [x] Build number is 0.
- [x] A tarball (`url`) rather than a repo (e.g. `git_url`) is used in your recipe (see [here](https://conda-forge.org/docs/maintainer/adding_pkgs.html#build-from-tarballs-not-repos) for more details).
- [x] GitHub users listed in the maintainer section have posted a comment confirming they are willing to be listed there.
- [x] When in trouble, please check our [knowledge base documentation](https://conda-forge.org/docs/maintainer/knowledge_base.html) before pinging a team.

### CODE REVIEW

File: recipes/ast-grep/meta.yaml [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+  name: {{ name|lower }}
+  version: {{ version }}
+
+source:
+  url: https://github.com/{{ name }}/{{ name }}/archive/refs/tags/{{ version }}.tar.gz
+  sha256: a1021a96fe778fcce7e600727e862b1b37f51031bad9f85d5d99c7f1aef2ffda
+
+build:
+  number: 0
+  skip: true # [win]
```
Comment (asford) [inline recipes/ast-grep/meta.yaml:14]:
```suggestion
  skip: true  # [win]
```

File: recipes/ast-grep/0001-Downgrade-tree-sitter-cpp-for-statis_assert-build-er.patch [hunk 2]
```diff
@@ -0,0 +1,49 @@
+From 8c566f0e4ae6533d73e300233814b8214a347819 Mon Sep 17 00:00:00 2001
+From: Alex Ford <alex.ford@abcellera.com>
+Date: Wed, 31 Jan 2024 16:03:24 -0800
+Subject: [PATCH] Downgrade tree-sitter-cpp for statis_assert build error
```
Comment (asford) [inline recipes/ast-grep/0001-Downgrade-tree-sitter-cpp-for-statis_assert-build-er.patch:4]:
```suggestion
Subject: [PATCH] Downgrade tree-sitter-cpp for static_assert build error
```

**Comment by asford** [comment]
@conda-forge/help-rust 

This feedstock is ready for review.

I need to make a changed to a pinned dependency due to a failure in static linking.
This is because the dependency has an improper use of `static_assert` for the C standard used in compilation of the feedstock.

This builds properly outside of the build container, so I suspect there may be some config needed to pass appropriate CFLAGS need to define the C standard version used during compilation? 
I think downgrading this dependency is acceptable for the time being.

File: recipes/ast-grep/meta.yaml [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+  skip: true  # [win]
+
+requirements:
+  build:
+    - {{ compiler('cxx') }}
+    - {{ compiler('rust') }}
+    - cargo-bundle-licenses
+  run:
+      # for details see conda_build_config.yaml
+    - __osx >=10.12     # [osx and x86_64]
```
Comment (xhochy) [inline recipes/ast-grep/meta.yaml:30]:
```suggestion
    - __osx >= {{ MACOSX_DEPLOYMENT_TARGET|default('10.12') }}     # [osx and x86_64]
```

File: recipes/ast-grep/meta.yaml [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+    write patterns as if you are writing ordinary code. It will match all code
+    that has the same syntactical structure. You can use $ sign + upper case
+    letters as wildcard, e.g. $MATCH, to match any single AST node. Think it as
+    REGEX dot ., except it is not textual.
+  license: MIT
+  license_family: MIT
+  # It is required to include a license file in the package,
+  # (even if the license doesn't require it) using the license_file entry.
+  # Please also note that some projects have multiple license files which all need to be added using a valid yaml list.
+  # See https://docs.conda.io/projects/conda-build/en/latest/resources/define-metadata.html#license-file
```
Comment (xhochy) [inline recipes/ast-grep/meta.yaml:52]:
```suggestion
```

File: recipes/ast-grep/meta.yaml [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+      # for details see conda_build_config.yaml
+    - __osx >=10.12     # [osx and x86_64]
+
+
+test:
+  commands:
+    - sg --version
+
+about:
+  home: https://github.com/simplejson/simplejson
```
Comment (xhochy) [inline recipes/ast-grep/meta.yaml:38]: This needs to be updated.

**Review by xhochy — CHANGES_REQUESTED** [review]
You could probably set a newer C std by appending it to the CFLAGS

File: recipes/ast-grep/meta.yaml (hunk 5 above)
Comment (asford) [reply recipes/ast-grep/meta.yaml:38]:
```suggestion
  home: https://github.com/ast-grep/ast-grep
```

File: recipes/ast-grep/meta.yaml [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+  skip: true  # [win]
+
+requirements:
+  build:
+    - {{ compiler('cxx') }}
+    - {{ compiler('rust') }}
+    - cargo-bundle-licenses
+  run:
+      # for details see conda_build_config.yaml
+    - __osx >= {{ MACOSX_DEPLOYMENT_TARGET|default('10.12') }}     # [osx and x86_64]
```
Comment (asford) [inline recipes/ast-grep/meta.yaml:30]:
```suggestion
    - __osx >={{ MACOSX_DEPLOYMENT_TARGET | default('10.12') }}     # [osx and x86_64]
```

**Comment by asford** [comment]
> You could probably set a newer C std by appending it to the CFLAGS

Thanks, I wasn't able to fully diagnose this yet but think we can roll with a build now with the downgraded dep.

**Review by xhochy — APPROVED** [review]

**Pull request merged**